Batch.Flush panics with `reflect.Value.Len on zero Value` when first column is all-NULL

**Describe the bug**

`Batch.Flush` panics when the **first argument position** across all `Add` calls is `nil` (NULL).

Two code paths combine to produce the panic:

`Batch.Add` — skips initialization for nil values:

```go
// batch.go, Add()
for i, v = range values {
    if !b.rValues[i].IsValid() {
        if v == nil { // a nil value has no type
            continue  // ← b.rValues[i] stays zero Value (invalid)
        }
        b.rValues[i] = reflect.MakeSlice(...)
    }
    // ...
}
```

When `v == nil` and `b.rValues[i]` has never been initialized, the loop skips that slot. This is correct for a *single* NULL in an otherwise typed column (a later non-nil row will initialize it). But when **every** row has NULL at position `i`, `b.rValues[i]` remains an invalid zero `reflect.Value`.

`Batch.Flush` — calls `.Len()` without `IsValid()` check:

```go
// batch.go:92, Flush()
if len(b.rValues) == 0 || b.rValues[0].Len() == 0 {
    return nil
}
```

`b.rValues[0]` is invalid → `.Len()` panics. The `Flush` loop below (which correctly handles invalid values by substituting `make([]string, b.size)`) is never reached.

**To Reproduce**

```go
package main

import (
	"context"
	"database/sql"
	"log"

	"github.com/godror/godror"
	_ "github.com/godror/godror"
)

func main() {
	db, err := sql.Open("godror", "user/pass@host/service")
	if err != nil {
		log.Fatal(err)
	}
	defer db.Close()

	ctx := context.Background()
	tx, err := db.BeginTx(ctx, nil)
	if err != nil {
		log.Fatal(err)
	}
	defer tx.Rollback()

	stmt, err := tx.PrepareContext(ctx,
		`INSERT INTO t (col_a, col_b) VALUES (:1, :2)`)
	if err != nil {
		log.Fatal(err)
	}
	defer stmt.Close()

	batch := &godror.Batch{Stmt: stmt}

	// First column is nil (NULL) — triggers the panic
	if err := batch.Add(ctx, nil, "hello"); err != nil {
		log.Fatal(err)
	}
	if err := batch.Add(ctx, nil, "world"); err != nil {
		log.Fatal(err)
	}

	// PANIC: reflect: call of reflect.Value.Len on zero Value
	if err := batch.Flush(ctx); err != nil {
		log.Fatal(err)
	}

	_ = tx.Commit()
}
```

DDL:
```sql
CREATE TABLE t (col_a VARCHAR2(100), col_b VARCHAR2(100));
```

**Expected behavior**

`Batch.Flush` should handle all-NULL columns gracefully, the same way it handles a single NULL row. The INSERT should succeed with NULLs in `col_a`.

**Error output**

```
panic: reflect: call of reflect.Value.Len on zero Value

goroutine 1 [running]:
reflect.Value.lenNonSlice(...)
	/usr/lib/golang/src/reflect/value.go:1782
reflect.Value.Len(...)
	/usr/lib/golang/src/reflect/value.go:1761
github.com/godror/godror.(*Batch).Flush(...)
	batch.go:92
```

**Your oracle client version**

23.3.0.0.0 (Oracle Instant Client)

**Your godror version**

v0.49.1 (also verified on v0.50.0 and `main` branch)

**Your go version**

go1.24

**Your gcc version**

Apple clang version 21.0.0 (clang-2100.1.1.101)

**Machine (please complete the following information):**
 - OS: Linux, macOS
 - Architecture: x86_64 (Linux), arm64 (macOS)

**Additional context**

Suggested fix — use `b.size` combined with `len(b.rValues)` to guard the early return instead of `b.rValues[0].Len()`:

```go
func (b *Batch) Flush(ctx context.Context) error {
    if b.size == 0 || len(b.rValues) == 0 {
        return nil
    }
    // ... rest of existing Flush logic
}
```

- `b.size == 0` covers the normal "no data to flush" path.
- `len(b.rValues) == 0` is a defensive guard for when `rValues` was never initialized (though unlikely under the current `Add` logic, it prevents future regressions).

The loop further down in `Flush` already handles invalid slots correctly:

```go
for i, v := range b.rValues {
    if !v.IsValid() {
        b.values[i] = make([]string, b.size) // empty string == NULL
    } else {
        b.values[i] = v.Interface()
    }
}
```

Once `Flush` gets past the early-return guard, the rest of the code works correctly with all-NULL columns. The issue is solely the guard at line 92.

Workaround — ensure the first `Add` call always has a typed value at every position. For an all-NULL column, pass `(*string)(nil)` (typed nil) instead of bare `nil`:

```go
var nullStr *string
batch.Add(ctx, nullStr, "hello") // b.rValues[0] gets initialized as []*string
```

--- a/batch.go
+++ b/batch.go
@@ -89,7 +89,7 @@ func (b *Batch) RowsAffected() int64 { return b.rowsAffected }
 
 // Flush executes the statement and clears the storage.
 func (b *Batch) Flush(ctx context.Context) error {
-	if len(b.rValues) == 0 || b.rValues[0].Len() == 0 {
+	if b.size == 0 || len(b.rValues) == 0 {
 		return nil
 	}
 

--- a/z_batch_test.go
+++ b/z_batch_test.go
@@ -616,3 +616,248 @@ func TestBatchFlushError(t *testing.T) {
 
 	t.Logf("Got expected error: %v", err)
 }
+
+// TestBatchAllNilColumn tests that Flush doesn't panic when a column is entirely nil (NULL) across all Add calls.
+// This verifies the fix for the issue #409 where rValues[0].Len() was called without IsValid() check.
+func TestBatchAllNilColumn(t *testing.T) {
+	t.Parallel()
+	ctx, cancel := context.WithTimeout(testContext("BatchAllNilColumn"), time.Minute)
+	defer cancel()
+
+	tbl := "test_batch_allnil" + tblSuffix
+	create := `CREATE TABLE ` + tbl + ` (col_a NUMBER(9), col_b NUMBER(9), col_c NUMBER(9))`
+	if _, err := testDb.ExecContext(ctx, create); err != nil {
+		t.Fatal(err)
+	}
+	defer func() {
+		_, _ = testDb.ExecContext(context.Background(), "DROP TABLE "+tbl)
+	}()
+
+	insQry := `INSERT INTO ` + tbl + ` (col_a, col_b, col_c) VALUES (:1, :2, :3)`
+	stmt, err := testDb.PrepareContext(ctx, insQry)
+	if err != nil {
+		t.Fatalf("%s: %+v", insQry, err)
+	}
+	defer stmt.Close()
+
+	b := godror.Batch{Stmt: stmt, Limit: 5}
+
+	// Add rows where col_a is nil in all calls - this should NOT panic on Flush
+	if err = b.Add(ctx, nil, 2, 3); err != nil {
+		t.Fatal(err)
+	}
+	if err = b.Add(ctx, nil, 4, 5); err != nil {
+		t.Fatal(err)
+	}
+	if err = b.Add(ctx, nil, 6, 7); err != nil {
+		t.Fatal(err)
+	}
+
+	// This was the panic point in the original bug - rValues[0].Len() on invalid Value
+	if err = b.Flush(ctx); err != nil {
+		t.Fatalf("Flush should not panic on all-nil column: %v", err)
+	}
+
+	// Verify rowsAffected count
+	if rowsAffected := b.RowsAffected(); rowsAffected != 3 {
+		t.Errorf("expected 3 rows affected, got %d", rowsAffected)
+	}
+
+	// Verify data was inserted correctly - NUMBER columns make NULL detection unambiguous
+	rows, err := testDb.QueryContext(ctx, "SELECT col_a, col_b, col_c FROM "+tbl+" ORDER BY col_b")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer rows.Close()
+
+	expected := []struct {
+		colA sql.NullInt64
+		colB int
+		colC int
+	}{
+		{sql.NullInt64{Valid: false}, 2, 3},
+		{sql.NullInt64{Valid: false}, 4, 5},
+		{sql.NullInt64{Valid: false}, 6, 7},
+	}
+
+	i := 0
+	for rows.Next() {
+		var colA sql.NullInt64
+		var colB, colC int
+		if err = rows.Scan(&colA, &colB, &colC); err != nil {
+			t.Fatal(err)
+		}
+		if i >= len(expected) {
+			t.Fatalf("more rows than expected: got row %d", i+1)
+		}
+		exp := expected[i]
+		if colA.Valid != exp.colA.Valid {
+			t.Errorf("row %d col_a: expected Valid=%v, got Valid=%v", i, exp.colA.Valid, colA.Valid)
+		}
+		if colB != exp.colB {
+			t.Errorf("row %d col_b: expected %d, got %d", i, exp.colB, colB)
+		}
+		if colC != exp.colC {
+			t.Errorf("row %d col_c: expected %d, got %d", i, exp.colC, colC)
+		}
+		i++
+	}
+	if i != len(expected) {
+		t.Errorf("expected %d rows, got %d", len(expected), i)
+	}
+}
+
+// TestBatchMultipleNilColumns tests multiple columns being entirely nil across all Add calls.
+func TestBatchMultipleNilColumns(t *testing.T) {
+	t.Parallel()
+	ctx, cancel := context.WithTimeout(testContext("BatchMultipleNilColumns"), time.Minute)
+	defer cancel()
+
+	tbl := "test_batch_multnil" + tblSuffix
+	create := `CREATE TABLE ` + tbl + ` (col_a NUMBER(9), col_b NUMBER(9), col_c NUMBER(9))`
+	if _, err := testDb.ExecContext(ctx, create); err != nil {
+		t.Fatal(err)
+	}
+	defer func() {
+		_, _ = testDb.ExecContext(context.Background(), "DROP TABLE "+tbl)
+	}()
+
+	insQry := `INSERT INTO ` + tbl + ` (col_a, col_b, col_c) VALUES (:1, :2, :3)`
+	stmt, err := testDb.PrepareContext(ctx, insQry)
+	if err != nil {
+		t.Fatalf("%s: %+v", insQry, err)
+	}
+	defer stmt.Close()
+
+	b := godror.Batch{Stmt: stmt, Limit: 5}
+
+	// Add rows where col_a and col_c are nil in all calls
+	if err = b.Add(ctx, nil, 10, nil); err != nil {
+		t.Fatal(err)
+	}
+	if err = b.Add(ctx, nil, 20, nil); err != nil {
+		t.Fatal(err)
+	}
+
+	// Should not panic when multiple columns are entirely nil
+	if err = b.Flush(ctx); err != nil {
+		t.Fatalf("Flush should not panic: %v", err)
+	}
+
+	// Verify all rows inserted - NUMBER columns make NULL detection unambiguous
+	rows, err := testDb.QueryContext(ctx, "SELECT col_a, col_b, col_c FROM "+tbl+" ORDER BY col_b")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer rows.Close()
+
+	expected := []struct {
+		colA sql.NullInt64
+		colB int
+		colC sql.NullInt64
+	}{
+		{sql.NullInt64{Valid: false}, 10, sql.NullInt64{Valid: false}},
+		{sql.NullInt64{Valid: false}, 20, sql.NullInt64{Valid: false}},
+	}
+
+	i := 0
+	for rows.Next() {
+		var colA, colC sql.NullInt64
+		var colB int
+		if err = rows.Scan(&colA, &colB, &colC); err != nil {
+			t.Fatal(err)
+		}
+		if i >= len(expected) {
+			t.Fatalf("more rows than expected: got row %d", i+1)
+		}
+		exp := expected[i]
+		if colA.Valid != exp.colA.Valid {
+			t.Errorf("row %d col_a: expected Valid=%v, got Valid=%v", i, exp.colA.Valid, colA.Valid)
+		}
+		if colB != exp.colB {
+			t.Errorf("row %d col_b: expected %d, got %d", i, exp.colB, colB)
+		}
+		if colC.Valid != exp.colC.Valid {
+			t.Errorf("row %d col_c: expected Valid=%v, got Valid=%v", i, exp.colC.Valid, colC.Valid)
+		}
+		i++
+	}
+	if i != len(expected) {
+		t.Errorf("expected %d rows, got %d", len(expected), i)
+	}
+}
+
+// TestBatchAllColumnsNil tests the extreme case where all columns are nil in all Add calls.
+func TestBatchAllColumnsNil(t *testing.T) {
+	t.Parallel()
+	ctx, cancel := context.WithTimeout(testContext("BatchAllColumnsNil"), time.Minute)
+	defer cancel()
+
+	tbl := "test_batch_allcolsnil" + tblSuffix
+	create := `CREATE TABLE ` + tbl + ` (col_a NUMBER(9), col_b NUMBER(9))`
+	if _, err := testDb.ExecContext(ctx, create); err != nil {
+		t.Fatal(err)
+	}
+	defer func() {
+		_, _ = testDb.ExecContext(context.Background(), "DROP TABLE "+tbl)
+	}()
+
+	insQry := `INSERT INTO ` + tbl + ` (col_a, col_b) VALUES (:1, :2)`
+	stmt, err := testDb.PrepareContext(ctx, insQry)
+	if err != nil {
+		t.Fatalf("%s: %+v", insQry, err)
+	}
+	defer stmt.Close()
+
+	b := godror.Batch{Stmt: stmt, Limit: 5}
+
+	// Add rows where ALL columns are nil
+	if err = b.Add(ctx, nil, nil); err != nil {
+		t.Fatal(err)
+	}
+	if err = b.Add(ctx, nil, nil); err != nil {
+		t.Fatal(err)
+	}
+
+	// Should not panic when all columns are nil - rValues[0] would be invalid
+	if err = b.Flush(ctx); err != nil {
+		t.Fatalf("Flush should not panic: %v", err)
+	}
+
+	// Verify rows were inserted as NULLs - NUMBER columns make NULL detection unambiguous
+	rows, err := testDb.QueryContext(ctx, "SELECT col_a, col_b FROM "+tbl)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer rows.Close()
+
+	expected := []struct {
+		colA sql.NullInt64
+		colB sql.NullInt64
+	}{
+		{sql.NullInt64{Valid: false}, sql.NullInt64{Valid: false}},
+		{sql.NullInt64{Valid: false}, sql.NullInt64{Valid: false}},
+	}
+
+	i := 0
+	for rows.Next() {
+		var colA, colB sql.NullInt64
+		if err = rows.Scan(&colA, &colB); err != nil {
+			t.Fatal(err)
+		}
+		if i >= len(expected) {
+			t.Fatalf("more rows than expected: got row %d", i+1)
+		}
+		exp := expected[i]
+		if colA.Valid != exp.colA.Valid {
+			t.Errorf("row %d col_a: expected Valid=%v, got Valid=%v", i, exp.colA.Valid, colA.Valid)
+		}
+		if colB.Valid != exp.colB.Valid {
+			t.Errorf("row %d col_b: expected Valid=%v, got Valid=%v", i, exp.colB.Valid, colB.Valid)
+		}
+		i++
+	}
+	if i != len(expected) {
+		t.Errorf("expected %d rows, got %d", len(expected), i)
+	}
+}

